mrmustard/lab_dev/states/base.py

```diff
@@ -648,6 +631,29 @@ def __init__(
         if representation is not None:
             self._representation = representation
 
+    def auto_shape(self) -> tuple[int, ...]:
+        r"""
+        A good enough estimate of the Fock shape of this DM, defined as the shape of the Fock
+        array (batch excluded) if it exists, and if it doesn't exist it is computed as the shape
+        that captures at least ``settings.AUTOCUTOFF_PROBABILITY`` of the probability mass of each
+        single-mode marginal (default 99.9%).
+        Note that it overwrites any ``None``s appearing in the ``fock_shape`` attribute and that
+        if a state is batched, the auto_shape is calculated for the first element.
+        """
+        try:  # fock
+            shape = self._representation.array.shape[1:]
+        except AttributeError:  # bargmann
+            repr = self.representation
+            shape = autoshape_numba(
+                math.asnumpy(repr.A[0]),
+                math.asnumpy(repr.b[0]),
+                math.asnumpy(repr.c[0]),
+            )
```

Comment:
Wasn't exactly sure where to put this comment, but I think we should run this on the renormalized state. For example, we may have a conditional state that is subnormalized, so it will never get to 99.9% and this will always send it to the max.

Example of a subnormalized state:
`(Coherent([0],0).dm() * 0.25).auto_shape() #shape (100,100)`

Reply:
More of an edge case (i.e. can't see how it would come about), for supernormalized states, this may truncate early.

Reply:
yes good idea

Reply:
done